Minio/Helper/RequestHelper.cs

```diff
@@ -0,0 +1,162 @@
+
+            // This is the actual url pointed to for all HTTP requests
+            endPoint = string.Format("{0}://{1}", scheme, host);
+            Uri uri = TryCreateUri(endPoint);
+            _validateEndpoint(uri);
+
+            return uri;
+        }
+        private static Uri TryCreateUri(string endpoint)
+        {
+            Uri uri = null;
+            try
+            {
+                uri = new Uri(endpoint);
+            }
+            catch (UriFormatException e)
+            {
+                throw new InvalidEndpointException(e.Message);
+            }
+            return uri;
+        }
+
+        /// <summary>
+        /// Validates URI to check if it is well formed. Otherwise cry foul.
+        /// </summary>
+        private static void _validateEndpoint(Uri uri)
+        {
+            if (string.IsNullOrEmpty(uri.OriginalString))
+            {
+                throw new InvalidEndpointException("Endpoint cannot be empty.");
+            }
+            string endpoint = uri.Host;
+
+            if (!isValidEndpoint(uri.Host))
+            {
+                throw new InvalidEndpointException(endpoint, "Invalid endpoint.");
+            }
+            if (!uri.AbsolutePath.Equals("/", StringComparison.CurrentCultureIgnoreCase))
+            {
+                throw new InvalidEndpointException(endpoint, "No path allowed in endpoint.");
+            }
+
+            if (!string.IsNullOrEmpty(uri.Query))
+            {
+                throw new InvalidEndpointException(endpoint, "No query parameter allowed in endpoint.");
+            }
+            if ((!uri.Scheme.ToLowerInvariant().Equals("https")) && (!uri.Scheme.ToLowerInvariant().Equals("http")))
+            //kp if (!(this.uri.Scheme.Equals(Uri.UriSchemeHttp) || this.uri.Scheme.Equals(Uri.UriSchemeHttps)))
+            {
+                throw new InvalidEndpointException(endpoint, "Invalid scheme detected in endpoint.");
+            }
+           
+        }
+
+        /// <summary>
+        /// Validate Url endpoint 
+        /// </summary>
+        /// <param name="endpoint"></param>
+        /// <returns>true/false</returns>
+        private static bool isValidEndpoint(string endpoint)
+        {
+            // endpoint may be a hostname
+            // refer https://en.wikipedia.org/wiki/Hostname#Restrictions_on_valid_host_names
+            // why checks are as shown below.
+            if (endpoint.Length < 1 || endpoint.Length > 253)
+            {
+                return false;
+            }
+
+            foreach (var label in endpoint.Split('.'))
+            {
+                if (label.Length < 1 || label.Length > 63)
+                {
+                    return false;
+                }
+
+                Regex validLabel = new Regex("^[a-zA-Z0-9][a-zA-Z0-9-]*");
+                Regex validEndpoint = new Regex(".*[a-zA-Z0-9]$");
+
+                if (!(validLabel.IsMatch(label) && validEndpoint.IsMatch(endpoint)))
```

Comment:
Seems like it should be `endpoint` instead of `label` in the above - or at least `endpoint` does not have to be checked in every iteration of the loop - it doesnt change. Also if the same string is being checked, the regular expressions can be concatenated and made into a single check.